### DISCUSSION TITLE
用Python的while迴圈(loop)連乘證明pi的數值

### DISCUSSION THREAD

**Discussion by enoch-lightning** [1101-GS4719-C/Python-Community-GS4719-C#202] [Q&A]
我使用Python的迴圈，搭配用紅色方框框起來的公式
<img width="960" alt="用來證明pi數值的連乘公式" src="https://user-images.githubusercontent.com/86505454/146474224-fe977a5e-5f6e-4050-838c-207ff3e35c30.png">
想要證明pi的數值有效至小數點以下7位(因為總共有整數位與小數點以下7位，所以是8位有效)為3.1415927。(「有效數字」在電腦的領域中好像也可以說是「精準度」？)但，使用google colab跑到最後，雖持續print出數值到電腦螢幕，但數值也不再改變。也發現下面的時間維持在2小時46分鐘34秒，沒有改變，旁邊還有紅色的!。
我不知道問題出在哪裡。我有將情況錄成影片，點選*用Python證明pi的數值*即可觀看。
[用Python證明pi的數值](https://youtu.be/FKijkntFPAM)
我發現2*x最後的結果停在3.141592643066262

**Comment by enoch-lightning** [comment]
這邊附上小寫pi與大寫pi的解釋。會連在一起，是因為我不知道要如何換行。
![image](https://user-images.githubusercontent.com/86505454/146479502-3e7dcb3a-406e-4257-a918-dfdbc9629b91.png)

**Comment by enoch-lightning** [comment]
今天(2021/12/18(六))我又嘗試一次，只是這次不印(print)出任何東西到瑩幕上。現在跑了11小時多都沒跑出來。
![image](https://user-images.githubusercontent.com/86505454/146647320-906d21cf-2760-499e-aecb-f6e22582fad6.png)

**Reply by enoch-lightning** [reply]
我把跑到第幾項(n)、跑了幾步(step)與2*()x跑到多少印出來(print)。*就是數學的x(乘以)。
![image](https://user-images.githubusercontent.com/86505454/146647495-515980d3-5a69-4f22-bdfc-0b63c36da888.png)

**Reply by enoch-lightning** [reply]
我發現這邊2*x的結果也是停在3.141592643066262。與跑兩個多小時的結果一樣。

**Reply by enoch-lightning** [reply]
跑兩個多小時(hr, h)，與跑十一點多小時，2*x的結果都一樣。可能原因第一應該是目前的資料型別只能儲存到小數點以下十五位。

**Reply by enoch-lightning** [reply]
跑這麼久都沒跑出來，假如題目沒出錯，應該就是迴圈邏輯正確，但不應該用迴圈寫。

**Reply by enoch-lightning** [reply]
我目前考慮使用numpy，證據如下
![image](https://user-images.githubusercontent.com/86505454/146955174-4e1ce69f-ac71-4990-a8d2-c2a263bcf175.png)
使用numpy處理的效率，是用python原生語法效率的(任務/12ms)/(任務/34.3s)=34.3s/12ms=34.3/12m=2858.3循環 倍
![image](https://user-images.githubusercontent.com/86505454/146955737-e9595c16-2bea-4fe5-af44-e603644292b9.png)
s=sec:秒，英文單數是second，英文複數是seconds

**Comment by Tomlord1122** [comment]
這應該只能算到一定的位數而已，因為電腦能負責的範圍是看有幾個bits，以C++ int來看他總共是4bytes ，也就是32bits
，其中一個bit拿來表示正負號，所以範圍是-2的31次方到正2的31次方減1。

**Comment by enoch-lightning** [comment]
我發現第三方函式庫numpy有函式(function)可以計算連乘，為prod()。但，發現有時會出現一個很奇怪的錯誤。
這個奇怪的錯誤如下
# numpy官網的奇怪錯誤範例
[https://numpy.org/doc/stable/reference/generated/numpy.prod.html#numpy.prod](https://numpy.org/doc/stable/reference/generated/numpy.prod.html#numpy.prod)
<img width="506" alt="numpy product()的奇怪錯誤" src="https://user-images.githubusercontent.com/86505454/147317746-a88f5ce9-b717-41fc-8166-b15e0202f792.png">

# 我在我的電腦的jupyter lab的測試
![image](https://user-images.githubusercontent.com/86505454/147317437-3f0d14a3-e4ac-4b4a-b631-19641187688c.png)

**問題：536870910的四次方怎麼會是16？**

**Comment by enoch-lightning** [comment]
我上numpy(寫NumPy也可以)的官網，查「文件(Documentation)」，查prod，發現有一大堆prod。我就產生了一個疑惑，這每一個prod都一樣嗎？
[NumPy的prod，網址如下
https://numpy.org/doc/stable/search.html?q=prod](https://numpy.org/doc/stable/search.html?q=prod)
![image](https://user-images.githubusercontent.com/86505454/147525613-67391b1a-a922-4dd8-b896-f9d1dfccbe43.png)
